pilot/pkg/config/aggregate/config.go

```diff
@@ -107,21 +108,33 @@ func (cr *store) Get(typ config.GroupVersionKind, name, namespace string) *confi
 
 // List all configs in the stores.
 func (cr *store) List(typ config.GroupVersionKind, namespace string) []config.Config {
-	if len(cr.stores[typ]) == 0 {
+	stores := cr.stores[typ]
+	if len(stores) == 0 {
 		return nil
 	}
-	var configs []config.Config
-	// Used to remove duplicated config
-	configMap := sets.New[types.NamespacedName]()
 
-	for _, store := range cr.stores[typ] {
-		storeConfigs := store.List(typ, namespace)
-		for _, cfg := range storeConfigs {
-			if !configMap.InsertContains(cfg.NamespacedName()) {
-				configs = append(configs, cfg)
-			}
-		}
+	var (
+		configs      []config.Config
+		storeConfigs = make([][]config.Config, 0, len(stores))
+		configCnt    int
+	)
+
+	for _, store := range stores {
+		curConfigs := store.List(typ, namespace)
+		storeConfigs = append(storeConfigs, curConfigs)
+		configCnt += len(curConfigs)
 	}
+
+	configs = make([]config.Config, 0, configCnt)
+	// Used to remove duplicated config
+	configMap := sets.NewWithLength[types.NamespacedName](configCnt)
+	for _, curConfigs := range storeConfigs {
+		configs = append(configs, curConfigs...)
+	}
+	configs = slices.FilterInPlace[config.Config](configs, func(cfg config.Config) bool {
+		return !configMap.InsertContains(cfg.NamespacedName())
```

Comment:
It is generally recommended to use stateless functions, but `configMap` is changed here.

Reply:
I understand that if this is a concurrent streaming process, stateful functions might pose concurrency issues. However, considering that this function is encapsulated by Istio and additional "concurrency" in Go is rare, this should not be a very strict constraint.

Alternatively, what would you suggest?

Reply:
I'm not sure, maybe like this

https://github.com/istio/istio/blob/705b68aba020229ba90b5032933461469b9a0846/pilot/pkg/networking/core/v1alpha3/httproute.go#L569

Reply:
u can see the original version https://github.com/istio/istio/compare/228d42007492798952db5b5bcd7b79f0fde5df51..fc706115770aed6055efdb876c0be8c2c673acaf .

Indeed, we have achieved cleaner and more readable code through some encapsulated syntactic sugar, but we have also lost some control over details. It's a trade-off. Following the suggestion above, I have switched to this version.

Reply:
Thanks, got it.

Reply:
IMO this is invalid. It relies on the assumption that Filter is called for each element in the slice once. This is probably true, but there is no guarantee. It should be stateless.

Why do we append then filter anyways? is it faster than filtering when we append like we did before?